Build failure on Cabal and Just grammars

### Summary

Getting build failures with the new Cabal (#6485) grammar and Just. Appears to be C++ related?

### Reproduction Steps

<!-- Ideally provide a key sequence and/or asciinema.org recording. --> 

I tried this:

1. `hx --grammar build`

I expected this to happen:

Grammars built with no errors

Instead, this happened:

Errors building `just` and `cabal` grammars.

<details><summary>grammar build log</summary>

```
Building 149 grammars
147 grammars already built
2 grammars failed to build
        Failure 0/2: Parser compilation failed.
Stdout:
Stderr: /Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:15:24: warning: default member initializer for non-static data member is a C++11 extension [-Wc++11-extensions]
  uint32_t prev_indent = 0;
                       ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:47:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto curr = std::to_string(state->prev_indent);
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:64:14: error: expected expression
    *state = {};
             ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:88:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto advance = [lexer] { lexer->advance(lexer, false); };
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:88:18: error: expected expression
  auto advance = [lexer] { lexer->advance(lexer, false); };
                 ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:89:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto skip = [lexer] { lexer->advance(lexer, true); };
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:89:15: error: expected expression
  auto skip = [lexer] { lexer->advance(lexer, true); };
              ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:91:3: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
  auto get_column = [lexer] { return lexer->get_column(lexer); };
  ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:91:21: error: expected expression
  auto get_column = [lexer] { return lexer->get_column(lexer); };
                    ^
/Users/bc/.config/helix/runtime/grammars/sources/just/src/scanner.cc:128:5: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
    auto indent = get_column();
    ^
6 warnings and 4 errors generated.

        Failure 1/2: Parser compilation failed.
Stdout:
Stderr: /Users/bc/.config/helix/runtime/grammars/sources/cabal/src/scanner.cc:26:60: error: expected expression
        for_each(indent_lvls.cbegin(), indent_lvls.cend(), [buffer, &i](uint8_t const n) { buffer[i++] = n; });
                                                           ^
/Users/bc/.config/helix/runtime/grammars/sources/cabal/src/scanner.cc:70:9: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
        auto cur_indent_lvl = indent_lvls.back();
        ^
1 warning and 1 error generated.
```
</details>

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
```

</details>


### Platform

macOS

### Terminal Emulator

Alacritty

### Helix Version

helix 23.03 (4cdba7cc)

--- a/helix-loader/Cargo.toml
+++ b/helix-loader/Cargo.toml
@@ -27,6 +27,7 @@ log = "0.4"
 # cloning/compiling tree-sitter grammars
 cc = { version = "1" }
 threadpool = { version = "1.0" }
+tempfile = "3.5.0"
 
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
 libloading = "0.7"

--- a/helix-loader/src/grammar.rs
+++ b/helix-loader/src/grammar.rs
@@ -8,6 +8,7 @@ use std::{
     process::Command,
     sync::mpsc::channel,
 };
+use tempfile::TempPath;
 use tree_sitter::Language;
 
 #[cfg(unix)]
@@ -446,13 +447,17 @@ fn build_tree_sitter_library(
         command.env(key, value);
     }
     command.args(compiler.args());
+    // used to delay dropping the temporary object file until after the compilation is complete
+    let _path_guard;
 
     if cfg!(all(windows, target_env = "msvc")) {
         command
             .args(["/nologo", "/LD", "/I"])
             .arg(header_path)
             .arg("/Od")
-            .arg("/utf-8");
+            .arg("/utf-8")
+            .arg("/std:c++14")
+            .arg("/std:c11");
         if let Some(scanner_path) = scanner_path.as_ref() {
             command.arg(scanner_path);
         }
@@ -466,20 +471,49 @@ fn build_tree_sitter_library(
             .arg("-shared")
             .arg("-fPIC")
             .arg("-fno-exceptions")
-            .arg("-g")
             .arg("-I")
             .arg(header_path)
             .arg("-o")
-            .arg(&library_path)
-            .arg("-O3");
+            .arg(&library_path);
+
         if let Some(scanner_path) = scanner_path.as_ref() {
             if scanner_path.extension() == Some("c".as_ref()) {
-                command.arg("-xc").arg("-std=c99").arg(scanner_path);
+                command.arg("-xc").arg("-std=c11").arg(scanner_path);
             } else {
-                command.arg(scanner_path);
+                let mut cpp_command = Command::new(compiler.path());
+                cpp_command.current_dir(src_path);
+                for (key, value) in compiler.env() {
+                    cpp_command.env(key, value);
+                }
+                cpp_command.args(compiler.args());
+                let object_file =
+                    library_path.with_file_name(format!("{}_scanner.o", &grammar.grammar_id));
+                cpp_command
+                    .arg("-fPIC")
+                    .arg("-fno-exceptions")
+                    .arg("-I")
+                    .arg(header_path)
+                    .arg("-o")
+                    .arg(&object_file)
+                    .arg("-std=c++14")
+                    .arg("-c")
+                    .arg(scanner_path);
+                let output = cpp_command
+                    .output()
+                    .context("Failed to execute C++ compiler")?;
+                if !output.status.success() {
+                    return Err(anyhow!(
+                        "Parser compilation failed.\nStdout: {}\nStderr: {}",
+                        String::from_utf8_lossy(&output.stdout),
+                        String::from_utf8_lossy(&output.stderr)
+                    ));
+                }
+
+                command.arg(&object_file);
+                _path_guard = TempPath::from_path(object_file);
             }
         }
-        command.arg("-xc").arg(parser_path);
+        command.arg("-xc").arg("-std=c11").arg(parser_path);
         if cfg!(all(
             unix,
             not(any(target_os = "macos", target_os = "illumos"))

